File: server/src/llm/lm_provider_googlegemini.py
from langchain_core.language_models.base import BaseLanguageModel
from langchain_core.embeddings.embeddings import Embeddings
from langchain_google_genai import GoogleGenerativeAIEmbeddings, ChatGoogleGenerativeAI
from aif_types.llm import LlmProvider
from llm.lm_base_provider import LmBaseProvider, LmBaseProviderProps


class LmProviderGoogleGemini(LmBaseProvider):
    def __init__(self):
        super().__init__(LmBaseProviderProps(
            id="googlegemini",
            name="Google Gemini",
            llmProvider=LlmProvider.GOOGLEGEMINI,
            jsonFileName="model_info/googlegemini_models.json",
            keyPrefix="GOOGLE_GEMINI_",
            apiKeyDescription="Doc: https://ai.google.dev/gemini-api/docs/api-key",
            apiKeyHint="",
        ))


    def _getBaseEmbeddingsModel(self, modelName: str, apiKey: str) -> Embeddings:
        return GoogleGenerativeAIEmbeddings(
            model=modelName,
            google_api_key=apiKey,
        )


    def _getBaseLanguageModel(self, modelName: str, apiKey: str) -> BaseLanguageModel:
        return ChatGoogleGenerativeAI(
            model=modelName,
            google_api_key=apiKey,
            temperature=0,
            streaming=True,
        )

File: packages/server-js/src/lm/LmProviderGoogleGemini.ts
import { Embeddings } from '@langchain/core/embeddings';
import { BaseChatModel } from "@langchain/core/language_models/chat_models";
import { ChatGoogleGenerativeAI, GoogleGenerativeAIEmbeddings } from "@langchain/google-genai";
import { AifUtils, api, LmProviderPropertyUtils } from 'aifoundry-vscode-shared';
import LmBaseProvider, { GetInitInfoResponse } from './LmBaseProvider';
import { HttpException } from '../exceptions';
import GoogleGeminiModels from "../config/model_info/googlegemini_models";
import LmProviderUtils from './LmProviderUtils';


enum CredPropKey {
    ApiKey = "ApiKey",
}

class LmProviderGoogleGemini extends LmBaseProvider {
    public static readonly ID = "googlegemini";

    protected async _getInitInfo(): Promise<GetInitInfoResponse> {
        const properties: Record<string, api.LmProviderProperty> = {
            [CredPropKey.ApiKey]: {
                description: "Google Gemini API key",
                hint: "",
                valueUri: null,
                isSecret: true,
            },
        }

        const modelMap = LmProviderUtils.createModelMap(GoogleGeminiModels, LmProviderGoogleGemini.ID);

        return {
            id: LmProviderGoogleGemini.ID,
            name: "Google Gemini",
            description: "",
            weight: 100,
            supportUserDefinedModels: false,
            modelMap,
            properties,
        };
    }

    public async isHealthy(): Promise<boolean> {
        const apiKey = this._databaseManager.getLmProviderInfo(this.id)?.properties[CredPropKey.ApiKey]?.valueUri as string;
        return !!apiKey && apiKey.length > 0;
    }

    public async getBaseEmbeddingsModel(aifUri: string): Promise<Embeddings> {
        const { apiKey, modelName } = await this._getCredentials(aifUri);
        return new GoogleGenerativeAIEmbeddings({
            apiKey,
            model: modelName,
            maxRetries: 1,
        });
    }

    public async getBaseLanguageModel(aifUri: string): Promise<BaseChatModel> {
        const { apiKey, modelName } = await this._getCredentials(aifUri);
        return new ChatGoogleGenerativeAI({
            apiKey,
            model: modelName,
        });
    }

    private async _getCredentials(aifUri: string): Promise<{
        apiKey: string,
        modelName: string,
    }> {
        const lmInfo = AifUtils.getModelNameAndVersion(this._info.id, aifUri);
        if (!lmInfo) {
            throw new HttpException(400, `Invalid uri ${aifUri}`);
        }

        const apiKey = await LmProviderPropertyUtils.getPropertyValue(this._info.properties[CredPropKey.ApiKey]);

        if (!apiKey) {
            throw new HttpException(400, "Invalid provider properties");
        }

        return {
            apiKey,
            modelName: lmInfo.modelName,
        };
    }
}

export default LmProviderGoogleGemini;
